app/src/main/java/ch/epfl/cs311/wanderwave/model/auth/AuthenticationController.kt

```diff
@@ -0,0 +1,55 @@
+package ch.epfl.cs311.wanderwave.model.auth
+
+import com.google.firebase.auth.FirebaseAuth
+import javax.inject.Inject
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.tasks.await
+
+class AuthenticationController @Inject constructor(private val auth: FirebaseAuth) {
+
+  fun isSignedIn(): Boolean {
+    return auth.currentUser != null
+  }
+
+  fun getUserData(): UserData? {
+    return auth.currentUser?.let { firebaseUser ->
+      UserData(
+          firebaseUser.uid,
+          firebaseUser.email,
+          firebaseUser.displayName,
+          firebaseUser.photoUrl?.toString())
+    }
+  }
+
+  fun authenticate(token: String): Flow<Boolean> {
+    if (auth.currentUser != null) {
+      return flowOf(true)
+    }
+    if (token.isEmpty()) {
+      return flowOf(false)
+    }
+    return flow {
+      val result = auth.signInAnonymously().await()
+      if (result.user != null) {
+        emit(true)
+      } else {
+        emit(false)
+      }
+    }
+  }
+
+  fun deauthenticate() {
+    auth.signOut()
+  }
+
+  private data class State(val isSignedIn: Boolean = false)
+
+  data class UserData(
+      val id: String,
+      val email: String?,
+      val displayName: String?,
+      val photoUrl: String?
+  )
```

Comment:
So you need this data class ? Why not use the profile class ? I know it's a bit different but it could be nice to all use the same one. We could add in the email. I can help you with that if you want. Just a question tho

Reply:
I do think it's a good idea to separate the parts relevant for authentication from our own data stored about a user. This one corresponds to the authentication data provided by firebase itself (later including Spotify), while the profile corresponds to our own objects in the firestore. But I definitely agree that we should connect them in a nice way. 
For now, the current user profile should have its firebaseUid set to `authenticationController.getUserData().id` to make sure permissions will always work, but with the current structure I'm unsure of where to implement that